File: 52. N-Queens II.go
package main

import (
	"fmt"
)

var counter int = 0

func backTrack(colSet, posDiagSet, negDiagSet map[int]bool, row, n int) {

	if row == n {
		counter++
		return
	}

	for col := 0; col < n; col++ {
		if colSet[col] || posDiagSet[col+row] || negDiagSet[row-col] {
			continue
		} else {
			colSet[col] = true
			posDiagSet[col+row] = true
			negDiagSet[row-col] = true
			backTrack(colSet, posDiagSet, negDiagSet, row+1, n)

			colSet[col] = false
			posDiagSet[col+row] = false
			negDiagSet[row-col] = false
		}
	}

}

func totalNQueens(n int) int {
	colSet := map[int]bool{}
	posDiagSet := map[int]bool{}
	negDiagSet := map[int]bool{}

	counter = 0

	backTrack(colSet, posDiagSet, negDiagSet, 0, n)

	fmt.Println("Counnter is: ", counter)

	return counter
}

func main() {
	fmt.Println("Hello World")
}

File: 52. N-Queens II.java
import java.util.HashSet;

class Solution {

    private HashSet<Integer> colSet = new HashSet<>();
    private HashSet<Integer> posDiagSet = new HashSet<>();
    private HashSet<Integer> negDiagSet = new HashSet<>();
    private Integer n;
    private Integer result = 0;

    private void backTracking(int row){

        if(row==n){
            System.out.println("addding: "+ row);
            result++;
            return;
        }

        for(int col=0;col<n;col++){

            if(colSet.contains(col) || posDiagSet.contains(col+row) || negDiagSet.contains(row-col)){
                continue;
            }

            colSet.add(col);
            posDiagSet.add(row+col);
            negDiagSet.add(row-col);

            backTracking(row+1);

            colSet.remove(col);
            posDiagSet.remove(row+col);
            negDiagSet.remove(row-col);

        }

    }

    public int totalNQueens(int n) {
        this.n = n;
        backTracking(0);
        return  result;
    }
}